psad.8
.\" Process this file with
.\" groff -man -Tascii foo.1
.\"
.TH PSAD 8 "March 2009" Linux
.SH NAME
.B psad
\- The Port Scan Attack Detector
.SH SYNOPSIS
.B psad [options]
.SH DESCRIPTION

.B psad
makes use of iptables log messages to detect, alert, and (optionally) block
port scans and other suspect traffic.  For TCP scans psad analyzes TCP
flags to determine the scan type (syn, fin, xmas, etc.) and corresponding
command line options that could be supplied to nmap to generate such a scan.
In addition, psad makes use of many TCP, UDP, and ICMP signatures contained
within the Snort intrusion detection system (see http://www.snort.org/) to
detect suspicious network traffic such as probes for common backdoors, DDoS
tools, OS fingerprinting attempts, and more.  By default psad also provides
alerts for snort rules that are detected directly by iptables through the
use of a ruleset generated by
.B fwsnort
(http://www.cipherdyne.org/fwsnort/).  This enables psad to send alerts for
application layer attacks.
.B psad
features a set of highly configurable danger thresholds (with sensible
defaults provided) that allow the administrator to define what constitutes
a port scan or other suspect traffic.  Email alerts sent by psad contain the
scanning ip, number of packets sent to each port, any TCP, UDP, or ICMP
signatures that have been matched (e.g. "NMAP XMAS scan"), the scanned port
range, the current danger level (from 1 to 5), reverse dns info, and whois
information.
.B psad
also makes use of various packet header fields associated with TCP SYN packets
to passively fingerprint remote operating systems (in a manner similar to the
.B p0f
fingerprinter) from which scans originate.  This requires the use of the
.B --log-tcp-options
argument for iptables logging rules; if this option is not used,
.B psad
will fall back to a fingerprinting method that makes use of packet length,
TTL and TOS values, IP ID, and TCP window sizes.
.PP
.B psad
reads all iptables log data by default from the
.I /var/log/messages
file.  By parsing firewall log messages, psad is provided with data that
represents packets that have been logged (and possibly dropped) by the
running iptables policy.  In this sense, psad is supplied with a pure data stream
that exclusively contains packets that the firewall has deemed unfit to
enter the network.
.B psad
consists of three daemons: psad, kmsgsd, and psadwatchd.
.B psad
is responsible for processing all packets that have been logged by the
firewall and applying the signature logic in order to determine what type
of scan has been leveraged against the machine and/or network.
.B kmsgsd
.B (deprecated)
reads all messages that have been written to the
.I /var/lib/psad/psadfifo
named pipe and writes any message that matches a particular regular
expression (or string) to
.I /var/log/psad/fwdata.
kmsgsd is only used if the
.B ENABLE_SYSLOG_FILE
variable is disabled in psad.conf.
.B psadwatchd
is a software watchdog that will restart any of the other two daemons should
a daemon die for any reason.
.SH OPTIONS
.TP
.BR \-A ", " \-\^\-Analyze-msgs
Analyze an iptables logfile for scans and exit.  This will generate email alerts
just as a normal running psad process would have for all logged scans.  By
default the psad data file
.I /var/log/psad/fwdata
is parsed for old scans, but any file can be specified through the use
of the \-\-messages-file command line option.  For example it might be useful
to point psad at your
.I /var/log/messages
file.
.TP
.BR \-\^\-analysis-fields\ \<search\ fields>
In \-\-Analyze mode restrict analysis to iptables log messages that have specific
values for particular fields.  Examples include "SRC:1.2.3.4", "DST:10.0.0.0/24,
and "TTL:64", and multiple fields are supported as a comma-separated list like
"SRC:1.2.3.4, LEN:44, DST:10.0.0.0/24".
.TP
.BR \-i "\fR,\fP " \-\^\-interface\ \<interface>
Specify the interface that
.B psad
will examine for iptables log messages.  This interface will be the
.B IN=
interface for packets that are logged in the
.B INPUT
and
.B FORWARD
chains, and the
.B OUT=
interface for packets logged in the
.B OUTPUT
chain.
.TP
.BR \-\^\-sig-update
Instruct
.B psad
to download the latest set of modified Snort signatures from
http://www.cipherdyne.org/psad/signatures so that psad can take advantage of
signature updates before a new release is made.
.TP
.BR \-O "\fR,\fP " \-\^\-Override-config\ \<file>
Override config variable values that are normally read from the
/etc/psad/psad.conf file with values from the specified file.  Multiple
override config files can be given as a comma separated list.
.TP
.BR \-D ", " \-\^\-Dump-conf
Dump the current psad config to STDOUT and exit.  Various pieces of information
such as the home network, alert email addresses, and DShield user id are removed
from the resulting output so it is safe to send to others.
.TP
.BR \-F ", " \-\^\-Flush
Remove any auto-generated firewall block rules if psad was configured
to automatically respond to scans (see the ENABLE_AUTO_IDS variable
in psad.conf).
.TP
.BR \-S ", " \-\^\-Status
Display the status of any psad processes that may or not be running.
The status output contains a listing of the number of packets that
have been processed by psad, along with all IP addresses and
corresponding danger levels that have scanned the network.
.TP
.BR \-\^\-status-ip\ \<ip>
Display status information associated with
.I ip
such as the protocol packet counters as well as the last 10 packets
logged by iptables.
.TP
.BR \-\^\-status-dl\ \<dl>
Display status information only for scans that have reached a danger
level of at least
.I dl
.TP
.BR \-\^\-status-summary
Instruct
.B psad
to omit detailed IP information from
.I --Status
and
.I --Analyze
modes.
.TP
.BR \-m "\fR,\fP " \-\^\-messages-file\ \<file>
This option is used to specify the file that will be parsed in analysis
mode (see the \-\-Analyze-msgs option).  The default path is the psad
data file
.I /var/log/psad/fwdata.
.TP
.BR \-\^\-CSV
Instruct
.B psad
to parse iptables log messages out of
.I /var/log/messages
(by defult, but this path can be changed with the
.I -m
option), and print the packet fields on STDOUT in comma-separate value
format.  This is useful for graphing iptables log data with AfterGlow
(see http://afterglow.sourceforge.net/index.html).
.TP
.BR \-\^\-stdin
Acquire iptables log data from STDIN instead of the default
.I /var/log/messages
file.
.TP
.BR \-\^\-CSV-fields\ \<tokens>
Instruct
.B psad
to only include a specific set of iptables log message fields within the CSV
output.  AfterGlow accepts up to three fields for its graph data, so the most
common usage of this option is "src dst dp" to print the source and destination
IP addresses, and the destination port number.
.TP
.BR \-K ", " \-\^\-Kill
Kill the current psad process along with psadwatchd and kmsgsd.
This provides a quick and easy way to kill all psad processes without
having to look in the process table or appeal to the psad-init script.
.TP
.BR \-R ", " \-\^\-Restart
Restart the currently running psad processes.  This option will
preserve the command line options that were supplied to the original
psad process.
.TP
.BR \-U ", " \-\^\-USR1
Send a running psad process a USR1 signal.  This will cause psad to
dump the contents of the %Scan hash to the file "/var/log/psad/scan_hash.$$"
where "$$" represents the pid of the psad process.  This is mostly
useful for debugging purposes, but it also allows the administrator to
peer into the %Scan hash, which is the primary data structure used to
store scan data within system memory.
.TP
.BR \-H ", " \-\^\-HUP
Send all running psad daemons a HUP signal.  This will instruct the
daemons to re-read their respective configuration files without causing
scan data to be lost in the process.
.TP
.BR \-B ", " \-\^\-Benchmark
Run psad in benchmark mode.  By default benchmark mode will simulate
a scan of 10,000 packets (see the \-\-packets option) and then report
the elapsed time.  This is useful to see how fast psad can process
packets on a specific machine.
.TP
.BR \-p "\fR,\fP " \-\^\-packets\ \<packets>
Specify the number of packets to analyze in \-\-Analyze mode or use in \-\-Benchmark mode.
The default is 10,000 packets in \-\-Benchmark mode, and unlimited in \-\-Analyze mode.
.TP
.BR \-d ", " \-\^\-debug
Run psad in debugging mode.  This will automatically prevent
psad from running as a daemon, and will print the contents
of the %Scan hash and a few other things on STDOUT at crucial
points as psad executes.
.TP

.BR \-c "\fR,\fP " \-\^\-config\ \<configuration-file>
By default all of the psad makes use of the configuration file
.I /etc/psad/psad.conf
for almost all configuration parameters.
.B psad
can be made to
override this path by specifying a different file on the command
line with the \-\-config option.
.TP
.BR \-\^\-signatures\ \<signatures-file>
The iptables firewalling code included within the linux 2.4.x kernel
series has the ability to distinguish and log any of the TCP flags
present within TCP packets that traverse the firewall interfaces.
.B psad
makes use of this logging capability to detect several types of TCP scan
signatures included within
.I /etc/psad/signatures.
The signatures were
originally included within the snort intrusion detection
system.  New signatures can be included and modifications to existing
signatures can be made to the signature file and psad will import
the changes upon receiving a HUP signal (see the \-\-HUP command line
option) without having to restart the psad process.
.B psad
also detects
many UDP and ICMP signatures that were originally included within snort.
.TP
.BR \-e ", " \-\^\-email-analysis
Send alert emails when run in \-\-Analyze-msgs mode.  Depending on the size of
the iptables logfile, using the \-\-email-analysis option could extend the runtime
of psad by quite a bit since normally both DNS and whois lookups will be issued
against each scanning IP address.  As usual these lookups can be disabled with
the \-\-no-rdns and \-\-no-whois options respectively.
.TP
.BR \-w ", " \-\^\-whois-analysis
By default
.B psad
does not issue whois lookups when running in \-\-Analyze-msgs mode.  The
\-\-whois-analysis option will override this behavior (when run in analysis mode)
and instruct psad to issue whois lookups against IP addresses from which scans
or other suspect traffic has originated.
.TP
.BR \-\^\-analysis-auto-block
Enable auto-blocking responses when running in \-\-Analyze-msgs mode.  This is
mostly useful only for the
.B psad
test suite when auto-blocking responses are tested and verified.
.TP
.BR \-\^\-snort-type\ \<type>
Restrict the type of snort sids to
.I type.
Allowed types match the file names given to snort rules files such as
"ddos", "backdoor", and "web-attacks".
.TP
.BR \-\^\-snort-rdir\ \<snort-rules-directory>
Manually specify the directory where the snort rules files are located.
The default is
.I /etc/psad/snort_rules.
.TP
.BR \-\^\-passive-os-sigs\ \<passive-os-sigs-file>
Manually specify the path to the passive operating system fingerprinting
signatures file.  The default is
.I /etc/psad/posf.
.TP
.BR \-\^\-auto-dl\ \<auto-dl-file>
Occasionally certain IP addresses are repeat offenders and
should automatically be given a higher danger level than
would normally be assigned.  Additionally, some IP addresses
can always be ignored depending on your network configuration
(the loopback interface 127.0.0.1 might be a good candidate
for example).
.I /etc/psad/auto_dl
provides an interface for psad to automatically
increase/decrease/ignore scanning IP danger levels.  Modifications
can be made to auto_dl (installed by default in /etc/psad)
and psad will import them with 'psad \-H' or by restarting the psad process.
.TP
.BR \-\^\-fw-search\ \<fw_search-file>
By default all of the psad makes use of the firewall search configuration
file
.I /etc/psad/fw_search.conf
for firewall search mode and search strings.
.B psad
can be made to
override this path by specifying a different file on the command
line with the \-\-fw-search option.
.TP
.BR \-\^\-fw-list-auto
List all rules in iptables chains that are used by
.B psad
in auto-blocking mode.
.TP
.BR \-\^\-fw-analyze
Analyze the local iptables ruleset, send any alerts if errors are
discovered, and then exit.
.TP
.BR \-\^\-fw-del-chains
By default, if ENABLE_AUTO_IDS is set to "Y"
.B psad
will not delete the auto-generated iptables chains (see the IPT_AUTO_CHAIN
keywords in psad.conf) if the \-\-Flush option is given.  The \-\-fw-del-chains
option overrides this behavior and deletes the auto-blocking chains from a
running iptables firewall.
.TP
.BR \-\^\-fw-dump
Instruct
.B psad
to dump the contents of the iptables policy that is running on the local
system.  All IP addresses are removed from the resulting output, so it is
safe to post to the psad list, or communicate to others.  This option is
most often used with \-\-Dump-conf.
.TP
.BR \-\^\-fw-block-ip\ \<ip>
Specify an IP address or network to add to the iptables controls that are
auto-generated by psad.  This allows psad to manage the rule timeouts.
.TP
.BR \-\^\-fw-rm-block-ip\ \<ip>
Specify an IP address or network to remove from the iptables controls that
are auto-generated by psad.
.TP
.BR \-\^\-fw-file\ \<policy-file>
Analyze the iptables ruleset contained within
.B policy-file
instead of the ruleset currently loaded on the local system.
.TP
.BR \-\^\-CSV-regex\ \<regex>
Instruct
.B psad
to only print CSV data that matches the supplied regex.  This regex is
used to match against each of the entire iptables log messages.
.TP
.BR \-\^\-CSV-neg-regex\ \<regex>
Instruct
.B psad
to only print CSV data that does not match the supplied regex.  This regex is
used to negatively match against each of the entire iptables log messages.
.TP
.BR \-\^\-CSV-uniq-lines
Instruct
.B psad
to only print unique CSV data.  That is, each line printed in
.I --CSV
mode will be unique.
.TP
.BR \-\^\-CSV-max-lines\ \<num>
Limit the number of CSV-formatted lines that
.B psad
generates on STDOUT.  This is useful to allow AfterGlow graphs to be
created that are not too cluttered.
.TP
.BR \-\^\-CSV-start-line\ \<num>
Specify the beginning line number to start parsing out of the iptables log
file in
.I --CSV
output mode.  This is useful for when the log file is extremely large, and you
want to begin parsing a specific place within the file.  The default is begin
parsing at the beginning of the file.
.TP
.BR \-\^\-CSV-end-line\ \<num>
Specify the ending line number to stop parsing the iptables log
file in
.I --CSV
output mode.  This is useful for when the log file is extremely large, and you
do not want
.B psad
to process the entire thing.
.TP
.BR \-\^\-gnuplot
Enter into Gnuplot mode whereby
.B psad
parses an iptables logfile and creates .gnu and .dat files that are suitable
for graphing with Gnuplot.  The various
.I --CSV
command line arguments apply to plotting iptables log with Gnuplot.
.TP
.BR \-\^\-gnuplot-template\ \<file>
Use a template file for all Gnuplot graphing directives (this is usually
a .gnu file by convention).  Normally
.B psad builds all of the graphing directives based on various --gnuplot
command line arguments, but the \-\-gnuplot-template switch allows you to
override this behavior.
.TP
.BR \-\^\-gnuplot-file-prefix\ \<file>
Specify a prefix for the .gnu, .dat, and .png files that are generated in
.I --gnuplot
mode.  So, when visualizing attacks captured in an iptables logfile (let's
say you are interested in port scans), you could use this option to have
.B psad
create the two files portscan.dat, portscan.gnu, and Gnuplot will create
an additional file portscan.png when the portscan.gnu file is loaded.
.TP
.BR \-\^\-gnuplot-x-label\ \<label>
Set the label associated with the x-axis.
.TP
.BR \-\^\-gnuplot-x-range\ \<range>
Set the x-axis range.
.TP
.BR \-\^\-gnuplot-y-label\ \<label>
Set the label associated with the y-axis.
.TP
.BR \-\^\-gnuplot-y-range\ \<range>
Set the y-axis range.
.TP
.BR \-\^\-gnuplot-z-label\ \<label>
Set the label associated with the z-axis (only if
.I --gnuplot-3D
is used).
.TP
.BR \-\^\-gnuplot-z-range\ \<range>
Set the z-axis range. (only if
.I --gnuplot-3D
is used).
.TP
.BR \-\^\-gnuplot-3D
Generate a Gnuplot
.B splot
graph.  This produces a three-dimensional graph.
.TP
.BR \-\^\-gnuplot-view
Set the viewing angle when graphing data in
.I --gnuplot-3D
mode.
.TP
.BR \-\^\-gnuplot-title\ \<title>
Set the graph title for the Gnuplot graph.
.TP
.BR \-I "\fR,\fP " \-\^\-Interval\ \<seconds>
Specify the interval (in seconds) that psad should use to
check whether or not packets have been logged by the
firewall.
.B psad
will use the default of 15 seconds unless a
different value is specified.
.TP
.BR \-l ", " \-\^\-log-server
This option should be used if psad is being executed on a syslog
logging server.  Running psad on a logging server requires that
check_firewall_rules() and auto_psad_response() not be executed
since the firewall is probably not being run locally.
.TP
.BR \-V ", " \-\^\-Version
Print the psad version and exit.
.TP
.BR \-\^\-no-daemon
Do not run psad as a daemon.  This option will display scan
alerts on STDOUT instead of emailing them out.
.TP
.BR \-\^\-no-ipt-errors
Occasionally iptables messages written by syslog to
.I /var/log/messages
seem to not conform to the normal firewall logging format if the kernel
ring buffer used by klogd becomes full.
.B
psad
will write these message to
.I /var/log/psad/errs/fwerrorlog
by default.  Passing the \-\-no-ipt-errors option will make psad ignore
all such erroneous firewall messages.
.TP
.BR \-\^\-no-whois
By default psad will issue a whois query against any IP from which
a scan has originated, but this can be disabled with the \-\-no-whois
command line argument.
.TP
.BR \-\^\-no-fwcheck
psad performs a rudimentary check of the firewall ruleset that
exists on the machine on which psad is deployed to determine
whether or not the firewall has a compatible configuration (i.e.
iptables has been configured to log packets).  Passing the
\-\-no-fwcheck or \-\-log-server options will disable this check.
.TP
.BR \-\^\-no-auto-dl
Disable auto danger level assignments.  This will instruct to not import
any IP addresses or networks from the file
.I /etc/psad/auto_dl.
.TP
.BR \-\^\-no-snort-sids
Disable snort sid processing mode.  This will instruct psad to not import
snort rules (for snort SID matching in a policy generated by
.B fwsnort
).
.TP
.BR \-\^\-no-signatures
Disable psad signature processing.  Note that this is independent of
snort SID matching in iptables messages generated by
.B fwsnort
and also from the ICMP type/code validation routines.
.TP
.BR \-\^\-no-icmp-types
Disable ICMP type and code field validation.
.TP
.BR \-\^\-no-passive-os
By default psad will attempt to passively (i.e. without sending
any packets) fingerprint the remote operating system from which
a scan originates.  Passing the \-\-no-passive-os option will
disable this feature.
.TP
.BR \-\^\-no-rdns
.B psad
normally attempts to find the name associated with a
scanning IP address, but this feature can be disabled with
the \-\-no-rdns command line argument.
.TP
.BR \-\^\-no-kmsgsd
Disable startup of kmsgsd.  This option is most useful for debugging
with individual iptables messages so that new messages are not appended
to the
.I /var/log/psad/fwdata
file.
.TP
.BR \-\^\-no-netstat
By default for iptables firewalls psad will determine whether
or not your machine is listening on a port for which a TCP
signature has been matched.  Specifying \-\-no-netstat
disables this feature.
.TP
.BR \-h ", " \-\^\-help
Print a page of usage information for psad and exit.
.SH FILES
.I /etc/psad/psad.conf
.RS
The main psad configuration file which contains configuration variables
mentioned in the section below.
.RE

.I /etc/psad/fw_search.conf
.RS
Used to configure the strategy both
.B psad
and
.B kmsgsd
employ to parse iptables messages.  Using configuration directive within
this file, psad can be configured to parse all iptables messages or only
those that match specific log prefix strings (see the \-\-log-prefix option
to iptables).
.RE

.I /etc/psad/signatures
.RS
Contains the signatures
.B psad
uses to recognize nasty traffic.  The
signatures are written in a manner similar to the *lib signature
files used in the snort IDS.
.RE

.I /etc/psad/icmp_types
.RS
Contains all valid ICMP types and corresponding codes as defined by RFC 792.
By default, ICMP packets are validated against these values and an alert
will be generated if a non-matching ICMP packet is logged by iptables.
.RE

.I /etc/psad/snort_rules/*.rules
.RS
Snort rules files that are consulted by default unless the \-\-no-snort-sids
commmand line argument is given.
.RE

.I /etc/psad/auto_dl
.RS
Contains a listing of any IP addresses that should be assigned
a danger level based on any traffic that is logged by the
firewall.  The syntax is "<IP address> <danger level>" where
<danger level> is an integer from 0 to 5, with 0 meaning to ignore
all traffic from <IP address>, and 5 is to assign the highest danger
level to <IP address>.
.RE

.I /etc/psad/posf
.RS
Contains a listing of all passive operating system fingerprinting
signatures.  These signatures include packet lengths, ttl, tos,
IP ID, and TCP window size values that are specific to various
operating systems.
.SH PSAD CONFIGURATION VARIABLES
This section describes what each of the more important
.B psad
configuration variables do and how they can be tuned to meet your
needs.  Most of the variables are located in the
.B psad
configuration file
.I /etc/psad/psad.conf
but the FW_SEARCH_ALL and FW_MSG_SEARCH variables are located in the
file
.I /etc/psad/fw_search.conf.
Each variable is assigned sensible defaults for most network
architectures during the install process.  More information on psad config
keywords may be found at:
.B http://www.cipherdyne.org/psad/config.html
.TP
.BR EMAIL_ADDRESSES
Contains a comma-separated list of email addresses to which email alerts
will be sent.  The default is "root@localhost".
.TP
.BR HOSTNAME
Defines the hostname of the machine on which
.B psad is running.  This will be
used in the email alerts generated by psad.
.TP
.BR HOME_NET
Define the internal network(s) that are connected to the local system.
This will be used in the signature matching code to determine whether traffic
matches snort rules, which invariably contain a source and destination
network.  Multiple networks are supported as a comma separated list, and
each network should be specified in CIDR notation.  Normally the network(s)

ChangeLog
psad-2.2.2 (//2013):
    - RPM bug fix to include the protocols file.

psad-2.2.1 (01/02/2013):
    - Added IP protocol scan detection (nmap -sO).  A new psad.conf variable
      PROTOCOL_SCAN_THRESHOLD defines the minimum number of different IP
      protocols (default = 5) that must be scanned before an alert is
      triggered.
    - Added detection for Topera IPv6 scans when --log-ip-options is used in
      the ip6tables logging rule.  When this option is not used, the previous
      psad-2.2 release detected Topera scans.  An example TCP SYN packet
      generated by Topera when --log-ip-options is used looks like this (note
      the series of empty IP options strings "OPT ( )":

        Dec 20 20:10:40 rohan kernel: [  488.495776] DROP IN=eth0 OUT=
        MAC=00:1b:b9:76:9c:e4:00:13:46:3a:41:36:86:dd
        SRC=2012:1234:1234:0000:0000:0000:0000:0001
        DST=2012:1234:1234:0000:0000:0000:0000:0002 LEN=132 TC=0 HOPLIMIT=64
        FLOWLBL=0 OPT ( ) OPT ( ) OPT ( ) OPT ( ) OPT ( ) OPT ( ) OPT ( )
        OPT ( ) OPT ( ) PROTO=TCP SPT=61287 DPT=1 WINDOW=8192 RES=0x00 SYN
        URGP=0

    - Bug fix in --Analyze mode when IP fields are to be searched with the
      --analysis-fields argument (such as --analysis-fields "SRC:1.2.3.4").
      The bug was reported by Gregorio Narvaez, and looked like this:

      Use of uninitialized value $_[0] in length at
      ../../blib/lib/NetAddr/IP/UtilPP.pm (autosplit into
      ../../blib/lib/auto/NetAddr/IP/UtilPP/hasbits.al) line 126.
      Use of uninitialized value $_[0] in length at
      ../../blib/lib/NetAddr/IP/UtilPP.pm (autosplit into
      ../../blib/lib/auto/NetAddr/IP/UtilPP/hasbits.al) line 126.
      Bad argument length for NetAddr::IP::UtilPP::hasbits, is 0, should be
      128 at ../../blib/lib/NetAddr/IP/UtilPP.pm (autosplit into
      ../../blib/lib/auto/NetAddr/IP/UtilPP/_deadlen.al) line 122.

    - Added --stdin argument to allow psad to collect iptables log data from
      STDIN in --Analyze mode.  This makes it easier to run an iptables logs
      through psad from arbitrary files like so:

      # grep "IN=.*OUT=" /var/log/kern.log | psad -A --stdin

    - Added the ability to acquire Snort rule 'msg' fields from fwsnort if
      it's also installed.  A new variable FWSNORT_RULES_DIR tells psad where
      to look for the fwsnort rule set.  This fixes a problem reported by Pui
      Edylie to the psad mailing list where fwsnort logged an attack that psad
      could not map back to a descriptive 'msg' field.
    - Added the ability to set per-danger level timeouts when psad is
      configured to run in auto-blocking mode.  These timeouts are implemented
      with new AUTO_BLOCK_DL*_TIMEOUT variables - one for each of the five
      possible danger levels that may be assigned to a scanning IP address.
    - Added the ability to throttle emails generated by psad via a new
      EMAIL_THROTTLE variable which is implemented as a per-IP threshold.  That
      is, if EMAIL_THROTTLE is set to "10", then psad will only send 1/10th as
      many emails for each scanning IP as it would have normally.

psad-2.2 (02/20/2012):
    - Added support for detection of malicious traffic that is delivered via
      IPv6.  This is accomplished by parsing ip6tables log messages - these are
      in a slightly different format than the iptables log messages.  Here is
      an example:

        Mar 17 13:39:13 linux kernel: [956932.483644] DROP IN=eth0 OUT=
        MAC=00:13:46:3a:41:36:00:1b:b9:76:9c:e4:86:dd
        SRC=2001:0db8:0000:f101:0000:0000:0000:0002
        DST=2001:0db8:0000:f101:0000:0000:0000:0001 LEN=80 TC=0 HOPLIMIT=64
        FLOWLBL=0 PROTO=TCP SPT=50326 DPT=993 WINDOW=5760 RES=0x00 SYN URGP=0

      Detection of malicious IPv6 traffic can be disabled via a new
      ENABLE_IPV6_DETECTION config variable.

    - For ICMP6 traffic, added protocol validation for ICMP6 type/code
      combinations.
    - Replaced Net::IPv4Addr with the excellent NetAddr::IP module which has
      comprehensive support for IPv6 address network parsing and comparisons.
    - Added a new test suite in the test/ directory to validate psad run time
      operations (scan detection, signature matching, and more).  To support
      this, a new '--install-test-dir' option was added to the install.pl
      script.  Once this is executed, the test suite can be run via the
      test-psad.pl script in the test/ directory.
    - Added a new MAX_SCAN_IP_PAIRS config variable to allow psad memory usage
      to be constrained by restricting the number of unique IP pairs that psad
      This is useful for when psad is deployed on systems with little memory,
      and is best utilized in conjunction with disabling ENABLE_PERSISTENCE so
      that old scans will also be deleted (and thereby making room for tracking
      new scans under the MAX_SCAN_IP_PAIRS threshold).
    - Bug fix for 'qw(...) usage as parenthesis' warnings for perl > 5.14
    - Bug fix that caused psad to emit the following:

        Undefined subroutine &main::LOG_DAEMON called at ./psad line 10071.

      This problem was noticed by Robert and reported on the psad mailing list.
    - Added --install-root to the install.pl script so that psad can be
      installed in a directory specified by the user as opposed to the normal
      system default.  This was a suggestion from @pyllyukko.
    - Added PERL5LIB env variable usage to the install.pl script so that module
      installs can reference the current install path.
    - Updated to the latest p0f signatures from OpenBSD.
    - Altered the 'ET MALWARE Bundleware Spyware CHM Download' Snort rule in
      the bundled Emerging Threats rule set to make sure that ClamAV does not
      flag on the pattern "mhtml\:file\://" which is associated with the
      following ClamAV signature:

        $ grep Exploit.HTML.MHTRedir-8 main.ndb
        Exploit.HTML.MHTRedir-8:3:*:6d68746d6c3a66696c653a2f2f{1-20}2168


Psad/lib/Psad.pm
#
##################################################################
#
# File: Psad.pm
#
# Purpose: This is the Psad.pm perl module.  It contains functions
#          that are reused by the various psad daemons.
#
# Author: Michael Rash (mbr@cipherdyne.org)
#
# Version: 2.0
#
##################################################################
#
# $Id$
#

package Psad;

use Exporter;
use Carp;
use strict;
use vars qw($VERSION @ISA @EXPORT @EXPORT_OK);

require Unix::Syslog;
Unix::Syslog->import(qw(:subs :macros));

require Exporter;

@ISA = qw(Exporter);
@EXPORT = qw(
    buildconf
    expand_vars
    defined_vars
    pidrunning
    writepid
    writecmdline
    unique_pid
    check_commands
    sendmail
    psyslog
    print_sys_msg
);

$VERSION = '2.0';

### subroutines ###
sub buildconf() {
    my ($config_hr, $cmds_hr, $conf_file) = @_;

    open C, "< $conf_file" or croak "[*] Could not open " .
        "config file $conf_file: $!";
    my @lines = <C>;
    close C;
    for my $line (@lines) {
        chomp $line;
        next if ($line =~ /^\s*#/);
        if ($line =~ /^\s*(\S+)\s+(.*?)\;/) {
            my $varname = $1;
            my $val     = $2;
            if ($val =~ m|/.+| && $varname =~ /^\s*(\S+)Cmd$/) {
                ### found a command
                $cmds_hr->{$1} = $val;
            } else {
                $config_hr->{$varname} = $val;
            }
        }
    }
    return;
}

sub expand_vars() {
    my ($config_hr, $cmds_hr) = @_;

    for my $hr ($config_hr, $cmds_hr) {
        for my $var (keys %$hr) {
            my $val = $hr->{$var};
            croak "[*] Multiple variable expansion not supported yet ",
                "(var $var)." if $val =~ m|\$.+\$|;
            if ($val =~ m|\$(\w+)|) {
                my $sub_var = $1;
                croak "[*] sub-ver $sub_var not allowed within same ",
                    "variable $var" if $sub_var eq $var;
                if (defined $config_hr->{$sub_var}) {
                    $val =~ s|\$$sub_var|$config_hr->{$sub_var}|;
                    $hr->{$var} = $val;
                } else {
                    croak "[*] sub-var \"$sub_var\" not defined in ",
                        "config for var: $var."
                }
            }
        }
    }
    return;
}

### check to make sure all required varables are defined in the config
### this subroutine is passed different variables by each script that
### correspond to only those variables needed be each script).
sub defined_vars() {
    my ($config_hr, $conf_file, $varnames_aref) = @_;
    for my $var (@$varnames_aref) {
        unless (defined $config_hr->{$var}) {  ### missing var
            croak "[*] The config file \"$conf_file\" does not " .
                  "contain the\nvariable: \"$var\".  Exiting!";
        }
    }
    return;
}

### check paths to commands and attempt to correct if any are wrong.
sub check_commands() {
    my ($cmds_hr, $exceptions_hr) = @_;
    my $caller = $0;
    my @path = qw(
        /bin
        /sbin
        /usr/bin
        /usr/sbin
        /usr/local/bin
        /usr/local/sbin
    );
    CMD: for my $cmd (keys %$cmds_hr) {
        ### both mail and sendmail are special cases, mail is not required
        ### if "nomail" is set in REPORT_METHOD, and sendmail is only
        ### required if DShield alerting is enabled and a DShield user
        ### email is set.
        next if $cmd =~ /mail/i;
        next if $cmd eq 'wget';  ### only used in --sig-update mode
        unless (-x $cmds_hr->{$cmd}) {
            my $found = 0;
            PATH: for my $dir (@path) {
                if (-x "${dir}/${cmd}") {
                    $cmds_hr->{$cmd} = "${dir}/${cmd}";
                    $found = 1;
                    last PATH;
                }
            }
            unless ($found) {
                unless (defined $exceptions_hr->{$cmd}) {
                    croak "[*] ($caller): Could not find $cmd ",
                        "anywhere!!!\n    Please edit the config section ",
                         "to include the path to $cmd.";
                }
            }
        }
        unless (-x $cmds_hr->{$cmd}) {
            unless (defined $exceptions_hr->{$cmd}) {
                croak "[*] ($caller): $cmd is located at ",
                    "$cmds_hr->{$cmd}, but is not executable\n",
                    "    by uid: $<";
            }
        }
    }
    return;
}

sub pidrunning() {
    my $pidfile = shift or croak '[*] Must supply a pid file.';
    return 0 unless -e $pidfile;
    open PIDFILE, "< $pidfile" or croak "[*] Could not open $pidfile: $!";
    my $pid = <PIDFILE>;
    close PIDFILE;
    chomp $pid;
    if (kill 0, $pid) {  ### pid is running
        return $pid;
    }
    return 0;
}

### make sure pid is unique
sub unique_pid() {
    my $pidfile = shift;
    croak "[*] $0 process is already running! Exiting.\n"
        if &pidrunning($pidfile);
    return;
}

### write the pid to the pid file
sub writepid() {
    my $pidfile = shift;
    my $caller = $0;
    open PIDFILE, "> $pidfile" or croak "[*] $caller: Could not " .
        "open pidfile $pidfile: $!\n";
    print PIDFILE $$ . "\n";
    close PIDFILE;
    chmod 0600, $pidfile;
    return;
}

### write command line to cmd file
sub writecmdline() {
    my ($args_aref, $cmdline_file) = @_;
    open CMD, "> $cmdline_file";
    print CMD "@$args_aref\n";
    close CMD;
    chmod 0600, $cmdline_file;
    return;
}

### send mail message to all addresses contained in the
### EMAIL_ADDRESSES variable within psad.conf ($addr_str).
### TODO:  Would it be better to use Net::SMTP here?
sub sendmail() {
    my ($subject, $body_file, $addr_str, $mailCmd) = @_;
    open MAIL, "| $mailCmd -s \"$subject\" $addr_str > /dev/null" or croak
        "[*] Could not send mail: $mailCmd -s \"$subject\" $addr_str: $!";
    if ($body_file) {
        open F, "< $body_file" or croak "[*] Could not open mail file: ",
            "$body_file: $!";
        my @lines = <F>;
        close F;
        print MAIL for @lines;
    }
    close MAIL;
    return;
}

### write a message to syslog
sub psyslog() {
    my ($ident, $msg) = @_;
    openlog($ident, &LOG_DAEMON(), &LOG_LOCAL7());
    syslog(&LOG_INFO(), $msg);
    closelog();
    return;
}

### write a message to a file
sub print_sys_msg() {
    my ($msg, $file) = @_;
    open F, ">> $file" or croak "[*] Could not open $file: $!";
    print F scalar localtime(), " pid: $$ $msg";
    close F;
    return;
}

1;
__END__

=head1 NAME

Psad - Perl extension for psad (the Port Scan Attack Detector) daemons

=head1 SYNOPSIS

  use Psad;
  writepid()
  writecmdline()
  unique_pid()
  psyslog()
  check_commands()

=head1 DESCRIPTION

The Psad.pm module contains several subroutines that are used by Port Scan
Attack Detector (psad) daemons.
writepid()  writes process ids to the pid files (e.g. "/var/run/psad.pid").
writecmdline()  writes the psad command line contained within @ARGV[] to the
file "/var/run/psad.cmd".
unique_pid()  makes sure that no other daemon process is already running in
order to guarantee pid uniqueness.
psyslog() an interface to sending messages via syslog.
check_commands()  check paths to commands and warns if unable to find any 
particular command.

=head1 AUTHOR

Michael Rash, mbr@cipherdyne.org

=head1 SEE ALSO

psad(8).
